pkg/controller/apiextensions/requirement/api.go

```diff
@@ -0,0 +1,120 @@
+*/
+
+package requirement
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+// Error strings.
+const (
+	errCreateComposite   = "cannot create composite resource"
+	errUpdateRequirement = "cannot update resource requirement"
+	errUpdateComposite   = "cannot update composite resource"
+	errDeleteComposite   = "cannot delete composite resource"
+)
+
+// An APICompositeCreator creates resources by submitting them to a Kubernetes
+// API server.
+type APICompositeCreator struct {
+	client client.Client
+	typer  runtime.ObjectTyper
+}
+
+// NewAPICompositeCreator returns a new APICompositeCreator.
+func NewAPICompositeCreator(c client.Client, t runtime.ObjectTyper) *APICompositeCreator {
+	return &APICompositeCreator{client: c, typer: t}
+}
+
+// TODO(negz): We should render and patch a composite resource on each
+// reconcile, rather than just creating it once.
+
+// Create the supplied composite using the supplied requirement.
+func (a *APICompositeCreator) Create(ctx context.Context, rq resource.Requirement, cp resource.Composite) error {
+
+	cp.SetRequirementReference(meta.ReferenceTo(rq, resource.MustGetKind(rq, a.typer)))
+	if err := a.client.Create(ctx, cp); err != nil {
+		return errors.Wrap(err, errCreateComposite)
+	}
+	// Since we use GenerateName feature of ObjectMeta, final name of the
+	// resource is calculated during the creation of the resource. So, we
+	// can generate a complete reference only after the creation.
+	cpr := meta.ReferenceTo(cp, resource.MustGetKind(cp, a.typer))
+	rq.SetResourceReference(cpr)
+
+	return errors.Wrap(a.client.Update(ctx, rq), errUpdateRequirement)
+}
+
+// An APIBinder binds requirements to composites by updating them in a
+// Kubernetes API server. Note that APIBinder does not support objects that do
+// not use the status subresource; such objects should use
+// APIBinder.
+type APIBinder struct {
+	client client.Client
+	typer  runtime.ObjectTyper
+}
+
+// NewAPIBinder returns a new APIBinder.
+func NewAPIBinder(c client.Client, t runtime.ObjectTyper) *APIBinder {
+	return &APIBinder{client: c, typer: t}
+}
+
+// Bind the supplied requirement to the supplied composite.
+func (a *APIBinder) Bind(ctx context.Context, rq resource.Requirement, cp resource.Composite) error {
+
+	cp.SetRequirementReference(meta.ReferenceTo(rq, resource.MustGetKind(rq, a.typer)))
+	if err := a.client.Update(ctx, cp); err != nil {
+		return errors.Wrap(err, errUpdateComposite)
+	}
+
+	if meta.GetExternalName(cp) == "" {
+		return nil
+	}
+
+	// Propagate back the final name of the composite resource to the requirement.
+	meta.SetExternalName(rq, meta.GetExternalName(cp))
+	return errors.Wrap(a.client.Update(ctx, rq), errUpdateRequirement)
```

Comment:
How about using `Apply` to patch?

Reply:
Typically we use Apply when we need to avoid resetting fields that we don't know about, right? Given this is all unstructured it's not obvious to me what benefit using `Apply` would get us.